src/modelfree/reward_shaping.py

```diff
@@ -0,0 +1,241 @@
+                    else:
+                        weighted_reward *= c
+                        self.dense_rews[env_num] += weighted_reward
+
+                if done[env_num] and self.sparse_rews[env_num] != 0:
+                    epinfo = {"sparse": self.sparse_rews[env_num],
+                              "dense": round(self.dense_rews[env_num], 2),
+                              "length": self.ep_lengths[env_num]}
+                    self._reset_stat_lists(idx=env_num)
+                    self.results_writer.write_row(epinfo)
+
+                shaped_reward += weighted_reward
+            rew[env_num] = shaped_reward
+
+        return obs, rew, done, infos
+
+    def get_annealed_exploration_reward(self):
+        """
+        Returns c (which will be annealed to zero) s.t. the total reward equals
+        c * reward_move + (1 - c) * reward_remaining
+        """
+        c = self.reward_annealer()
+        assert 0 <= c <= 1
+        return c
+
+
+class NoisyAgentWrapper(ResettableAgent):
+    def __init__(self, agent, noise_annealer, noise_type='gaussian'):
+        """
+        - agent: ResettableAgent (most likely) - noise will be added to the actions of
+        this agent in order to build a curriculum of weaker to stronger victims
+        - noise_annealer: Annealer.get_value - presumably the noise should be decreased over time
+        in order to get the adversarial policy to perform well on a normal victim.
+        This function should be tied to a Scheduler to keep it stateless.
+        - noise_type: str - the type of noise parametrized by noise_annealer's value.
+        Current options are [gaussian]
+        """
+        self.agent = agent
+        self.noise_annealer = noise_annealer
+        self.noise_generator = self._get_noise_generator(noise_type)
+
+    def _get_noise_generator(self, noise_type):
+        noise_generators = {
+            'gaussian': lambda x, size: np.random.normal(scale=x, size=size)
+        }
+        return noise_generators[noise_type]
+
+    def get_action(self, observation):
+        noise_param = self.noise_annealer()
+        original_action = self.agent.get_action(observation)
+        action_size = original_action.shape
+
+        noise = self.noise_generator(noise_param, action_size)
+        noisy_action = original_action + noise
+        return noisy_action
+
+    def reset(self):
+        return self.agent.reset()
+
+
+class HumanoidEnvWrapper(VecEnvWrapper):
+    def __init__(self, env):
+        super().__init__(env)
+        self.env = env
+
+    def reset(self):
+        return self.env.reset()
+
+    def step_wait(self):
+        obs, rew, done, infos = self.env.step_wait()
+        num_envs = len(obs)
+        for env_num in range(num_envs):
+            rew[env_num] -= infos[env_num]['reward_linvel']
+
+        return obs, rew, done, infos
+
+
+# TODO: should this inherit from something?
+class Scheduler(object):
```

Comment:
There's a lot of code duplication here. You're doing basically the same thing for {get,set}_{lr,rew,noise}_{func,val}. Can you refactor this to have a common interface?

I'm thinking take a dict of callables, and then just have a get_val(k, frac_remaining) function which: (a) updates frac remaining if non-None, (b) calls func[k](self.frac_remaining). You can also have a `set_func(k, f)`, although I'm not sure whether this is used.